Docs for conflicts (#72)
aerogear/voyager-server: doc/guides/conflicts.md

## Conflict resolution

## Prerequisites:

- GraphQL server with resolvers
- Database or any other form of data storage that can cause data conflicts

## Pluggable conflict resolution

Pluggable conflict resolution is an concept that allows developers to define handling for collisions between data sent to resolver

> and data that is currently stored in the server.

Reviewer: I know markdown will fix it, but it should not have a breaking line

Reviewer 2: @danielpassos I wouldn't be too worried about this because usually it makes the file more readable inside an editor 🙂

In fact a lot of projects actually enforce max line lengths in markdown files and force you to do this. I would personally never go that far and enforce it but I so generally see this as a good thing!

Pluggable conflict resolution supports following implementations:
- `VersionedObjectState`: depends on version field supplied in objects (used by default when importing conflictHandler)
- `HashObjectState`: depends on hash calculated from entire object

## Enabling conflict resolution mechanism

To enable conflict resolution developers need to use one of the pluggable conflict resolution strategies
in each individual resolver. Depending on strategy developers will need to provide additional details.

### Version based conflict resolution

1. Add conflict package dependency to project

```javascript
const { conflictHandler } = require('@aerogear/apollo-voyager-conflicts')
```

2. Add version field to GraphQL type that should support conflict resolution

```graphql
type Greeting {
    msg: String
    version: Int
  }
```

3. Add example mutations

```graphql
  type Mutation {
    changeGreeting(msg: String!, version: Int!): Greeting
  }
```

> 3. Implement resolver for mutation

Reviewer: Should be 4

Every conflict can be handled using set of predefined steps

```javascript
    // 1. Read data from data source
    // 2. Check for conflicts
    if (conflictHandler.hasConflict(serverData,clientData)) {
      // 3. Resolve conflict (client or server) and return error to client
      return await conflictHandler.resolveOnClient(serverData, clientData).response
    }

>     // 5. Call next state to update

Reviewer: Where is the step 4?

Author: Actually rendering works fine that's why I missed that.

> ## Conflict resolution

Reviewer 2: ```suggestion
## Conflict Resolution
```

## Prerequisites:

- GraphQL server with resolvers
- Database or any other form of data storage that can cause data conflicts

> ## Pluggable conflict resolution

Reviewer 2: Minor thing but headings should always be captialized with the exception of words like `the, and, of, for, in, at` etc. (unless they are at the beginning of the heading of course).

```suggestion
## Pluggable Conflict Resolution
```

> Pluggable conflict resolution is an concept that allows developers to define handling for collisions between data sent to resolver

Reviewer 3: ```suggestion
Pluggable conflict resolution is a concept that allows developers to define handling for collisions between data sent to resolver
```

and data that is currently stored in the server.

> Pluggable conflict resolution supports following implementations:

Reviewer 3: ```suggestion
Pluggable conflict resolution supports the following implementations:
```

- `VersionedObjectState`: depends on version field supplied in objects (used by default when importing conflictHandler)
- `HashObjectState`: depends on hash calculated from entire object

> Both implementations basing on `ObjectState` interface that can be extended to provide another custom implementation for conflict detection

Reviewer 3: ```suggestion
Both implementations are based on the `ObjectState` interface that can be extended to provide another custom implementation for conflict detection
```

## Enabling conflict resolution mechanism

To enable conflict resolution developers need to use one of the pluggable conflict resolution strategies

> in each individual resolver. Depending on strategy developers will need to provide additional details.

Reviewer 3: ```suggestion
in each individual resolver. Depending on the strategy developers will need to provide additional details.
```

### Version based conflict resolution

1. Add conflict package dependency to project

```javascript
const { conflictHandler } = require('@aerogear/apollo-voyager-conflicts')
```

2. Add version field to GraphQL type that should support conflict resolution

```graphql
type Greeting {
    msg: String
    version: Int
  }
```

3. Add example mutations

```graphql
  type Mutation {
    changeGreeting(msg: String!, version: Int!): Greeting
  }
```

4. Implement resolver for mutation

> Every conflict can be handled using set of predefined steps

Reviewer 3: ```suggestion
Every conflict can be handled using a set of predefined steps
```

```javascript
    // 1. Read data from data source
    // 2. Check for conflicts
    if (conflictHandler.hasConflict(serverData,clientData)) {
      // 3. Resolve conflict (client or server) and return error to client
      return await conflictHandler.resolveOnClient(serverData, clientData).response
    }
    // 4. Call next state to update
    greeting = conflictHandler.nextState(clientData)
    // 5. Save object to data source
```

> Resolvers can be implemented to handle conflicts on client or server.

Reviewer 3: ```suggestion
Resolvers can be implemented to handle conflicts on the client or the server.
```

> Depending on handling resolver implementation will differ.

Reviewer 3: ```suggestion
Depending on the strategy used, the resolver implementation will differ.
```

> Please see chapter bellow for individual implementations.

Reviewer 2: ```suggestion
Please see the chapter below for individual implementations.
```

// Data
const serverState = ...

changeGreeting: async (obj, clientState, context, info) => {
    if (conflictHandler.hasConflict(serverState, args)) {
      const clientState = args
      return await conflictHandler.resolveOnClient(serverState, clientState).response
    }
    serverState = conflictHandler.nextState(clientState)
    return serverState
}
```

### Resolving conflicts on the server

```javascript
// Data
const serverState = ...

 changeGreeting: async (obj, clientState, context, info) => {
      if (conflictHandler.hasConflict(serverState, clientState)) {
        const strategy = customGreetingResolutionStrategy
        const { resolvedState, response } = await conflictHandler.resolveOnServer(strategy, serverState, clientState)
        serverState = resolvedState
        return response
      }
      serverState = conflictHandler.nextState(clientState)
      return serverState
    }
```

> Note: For complete implementation see example application located in `examples/conflicts` folder.

## Client conflict implementation

See [Voyager Client documentation](https://github.com/aerogear/aerogear-js-sdk/tree/master/packages/sync#conflicts)

## `ObjectState` interface

> `ObjectState` interface is an complete conflict resolution implementation that provides set of rules to detect and handle conflict. Interface will allow developers to handle conflict on client or server. `nextSate` method is a way for interface to modify existing object before is being saved to the database.

Reviewer 3: ```suggestion
The `ObjectState` interface is a complete conflict resolution implementation that provides a set of rules to detect and handle conflict. This interface will allow developers to handle a conflict on the client or the server. The `nextSate` method is a way for the interface to modify the existing object before it is saved to the database.
For example when using the `lastModified` field as a way to detect conflicts:
```
